pkg/common/cns-lib/volume/util.go

```diff
@@ -337,9 +342,22 @@ func getCnsVolumeInfoFromTaskResult(ctx context.Context, virtualCenter *cnsvsphe
 	}, "", nil
 }
 
+// classifyAsNonStorageFault checks if the fault type is in a pre-defined list of non-storage faults
+// and returns a bool value accordingly.
+func classifyAsNonStorageFault(fault string) bool {
+	for _, nonStorageFault := range VimNonStorageFaultsList {
+		if nonStorageFault == fault {
+			return true
+		}
+	}
+	return false
+}
+
 // ExtractFaultTypeFromErr extracts the fault type from err.
-// Return the vim fault type if the input err is a SoapFault, and can exract the fault type of VimFault.
-// Otherwise, it returns fault type as "csi.fault.Internal".
+// If the input err is a SoapFault and fault type of VimFault can be extracted, then
+// a. Return vim fault type if fault originates in downstream storage components.
+// b. Return nonstorage fault type if it's known to originate in other vSphere components.
+// Return "csi.fault.Internal" fault type if the input err is not a Soap fault.
 func ExtractFaultTypeFromErr(ctx context.Context, err error) string {
```

Comment:
It looks like the intention behind writing this code was to expose the underlying fault reported by vCenter. But we are changing those semantics by customizing some vCenter faults to CSI-level non-storage faults. Can we get this reviewed by @divyenpatel ?

Just wondering - can we not handle the problem at the place where we are reporting the metrics? For example, before setting the fault type in the Prometheus metric, we could convert the known non-storage vim faults to non-storage tags and report them in the metric. For unknown non-storage faults, they will get reported as is.

Reply:
>can we not handle the problem at the place where we are reporting the metrics? 

That's also one way to handle it. But I think the whole rationale behind introducing methods like `ExtractFaultTypeFromErr` & `ExtractFaultTypeFromVolumeResponseResult` was to propagate the fault type back to caller so that they can eventually be used for `faultType` label in Prometheus metrics. There's no other use case for this. I think it's better to handle this at one place rather than handling in each API separately.

Reply:
The methods were originally introduced as part of this MR - https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/1219.  
@divyenpatel & @lipingxue can also weigh in on the approach in this MR.

Reply:
so far returned `faulttype` from the func `ExtractFaultTypeFromErr` is only used by Prometheus metrics.
but InvalidHostState fault is derived from InvalidState fault which is derived from VimFault so technically InvalidHostState is also vimfault.

can we not add another function to classify non storage vim faults and keep  `ExtractFaultTypeFromErr` as it is.

Reply:
We're adding the prefix `vim.fault.` to fault string in `ExtractFaultTypeFromErr` method. For instance, `types. InvalidHostState` from CNS becomes `vim.fault.InvalidHostState` in this method and that's what propagated to Prometheus.  
Ideally we want to return `csi.fault.nonstorage.X` for non-storage faults.  Should we also preserve `vim.fault` keyword in the fault to indicate it's a vim fault? If we want to preserve `vim.fault.` keyword then it will become `csi.fault.nonstorage.vim.fault.X` and can be handled in a separate function. If not, we'd be first adding the prefix here and then replacing it in another method which doesn't make sense.

Reply:
Can't we catch `vim.fault.InvalidHostState` and just return fault type as nonstorage with the fault as `vim.fault.InvalidHostState`?

Reply:
@divyenpatel As discussed offline, I have kept `ExtractFaultTypeFromErr` unchanged and added `csi.fault.nonstorage.` prefix to vim faults in the latest commit.